### PR TITLE
Support for SUSE/openSUSE ipaplatform

### PR DESCRIPTION
**Pull request by hellcp**
Upstreaming much of the work done for FreeIPA 4.6 and updated for 4.8 (well, master)

Building the snapshots of this over on https://build.opensuse.org/package/show/home:hellcp/freeipa, although the build service workers time out while building python parts, so not everything might build there until I tweak the spec (or workers) ;)

### CODE REVIEW

**Comment by tiran** [comment]
Thanks for your contribution! It's good to finally see SuSE support landing in upstream.

The new code uses a slightly different code style than the rest of our code (e.g. single quotes). Could you please run [``black -l78``](https://pypi.org/project/black/) to auto-format the new code.

File: ipaplatform/suse/paths.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    KDC_CERT = '/var/lib/kerberos/krb5kdc/kdc.crt'
+    KDC_KEY = '/var/lib/kerberos/krb5kdc/kdc.key'
+    NAMED_RUN = '/var/lib/named/data/named.run'
+    IPA_CUSTODIA_HANDLER = '/usr/lib/ipa/custodia'
+    WSGI_PREFIX_DIR = '/run/apache2/wsgi'
+    KDESTROY = '/usr/lib/mit/bin/kdestroy'
+    BIN_KVNO = '/usr/lib/mit/bin/kvno'
+    UPDATE_CA_TRUST = '/usr/sbin/update-ca-certificates'
+    AUTHSELECT = '/usr/bin/authselect'
+    VAR_RUN_DIRSRV_DIR = '/run/dirsrv'
```
Comment (tiran) [inline ipaplatform/suse/paths.py:100]: You can remove most ``/run`` overrides. I have created https://github.com/freeipa/freeipa/pull/4533 to fix the base paths.

File: ipaplatform/suse/services.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    system_units = suse_system_units
+
+    def __init__(self, service_name, api=None):
+        systemd_name = service_name
+        if service_name in self.system_units:
+            systemd_name = self.system_units[service_name]
+        else:
+            if '.' not in service_name:
+                systemd_name = "%s.service" % (service_name)
+        super(SuseService, self).__init__(service_name, systemd_name, api)
```
Comment (tiran) [inline ipaplatform/suse/services.py:62]: Let's make new code use proper Python 3 syntaxy:
``super().__init__(service_name, systemd_name, api)``

File: ipaplatform/suse/services.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+                instance_name, capture_output=capture_output, wait=wait)
+
+    def start(self, instance_name="", capture_output=True, wait=True,
+              ldapi=False):
+        with self.__wait(instance_name, wait, ldapi) as wait:
+            super(SuseDirectoryService, self).start(
+                instance_name, capture_output=capture_output, wait=wait)
+
+    @contextlib.contextmanager
+    def __wait(self, instance_name, wait, ldapi):
```
Comment (tiran) [inline ipaplatform/suse/services.py:109]: Please use single underscore for private functions, ``def _wait``

File: ipaplatform/suse/services.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            ipautil.wait_for_open_socket(socket_name,
+                                         self.api.env.startup_timeout)
+        else:
+            yield wait
+
+
+class SuseIPAService(SuseService):
+    # Credits to upstream developer
+    def enable(self, instance_name=""):
+        super(SuseIPAService, self).enable(instance_name)
```
Comment (tiran) [inline ipaplatform/suse/services.py:132]: ``super()``

File: ipaplatform/suse/services.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+    # Credits to upstream developer
+    def __init__(self):
+        # pylint: disable=ipa-forbidden-import
+        import ipalib
+        # pylint: enable=ipa-forbidden-import
+        services = dict()
+        for s in base_services.wellknownservices:
+            services[s] = self.service_class_factory(s, ipalib.api)
+        super(SuseServices, self).__init__(services)
```
Comment (tiran) [inline ipaplatform/suse/services.py:196]: ``super()``

**Comment by tiran** [comment]
The copyright header is
```
#
# Copyright (C) 2020 FreeIPA Contributors see COPYING for license
#
```
with two empty lines before/after.

Could you please squash your commits into a single commit to reduce noise?

**Comment by abbra** [comment]
LGTM

**Comment by tiran** [comment]
master:

* 3e8c51922b404bf2406ead09376e3e89f0624cdf Support for SUSE/openSUSE ipaplatform

**Pull request closed without merging**